mqtt/mqtt-bridge/src/pump.rs

```diff
@@ -0,0 +1,170 @@
+        let publish_handle = client
+            .publish_handle()
+            .map_err(BridgeError::PublishHandle)?;
+        let client_shutdown = client.shutdown_handle()?;
+
+        Ok(Self {
+            client,
+            client_shutdown,
+            publish_handle,
+            subscriptions,
+            loader,
+            persist,
+            pump_type,
+            bridge_name,
+        })
+    }
+
+    pub async fn subscribe(&mut self) -> Result<(), BridgeError> {
+        self.client
+            .subscribe(&self.subscriptions)
+            .await
+            .map_err(BridgeError::Subscribe)?;
+
+        Ok(())
+    }
+
+    pub async fn run(&mut self, shutdown: Receiver<()>) {
+        debug!("starting pumps for {} bridge...", self.bridge_name);
+
+        let (loader_shutdown, loader_shutdown_rx) = oneshot::channel::<()>();
+        let publish_handle = self.publish_handle.clone();
+        let persist = self.persist.clone();
+        let loader = self.loader.clone();
+        let mut client_shutdown = self.client_shutdown.clone();
+        let bridge_name = self.bridge_name.clone();
+        let pump_type = self.pump_type.clone();
+
+        // egress pump
+        let f1 = async move {
+            // TODO REVIEW: Is panicking fine?
+            let mut loader_borrow = loader.borrow_mut();
+            let mut receive_fut = loader_shutdown_rx.into_stream();
+
+            debug!(
+                "{} bridge starting egress publication processing for {:?} pump...",
+                bridge_name, pump_type
+            );
+
+            loop {
+                let mut publish_handle = publish_handle.clone();
+                match select(receive_fut.next(), loader_borrow.try_next()).await {
+                    Either::Left((shutdown, _)) => {
+                        debug!("egress pump received shutdown signal");
+                        if let None = shutdown {
+                            error!(message = "unexpected behavior from shutdown signal while signalling bridge pump shutdown")
+                        }
+
+                        debug!("bridge pump stopped");
+                        break;
+                    }
+                    Either::Right((loaded_element, _)) => {
+                        debug!("egress pump extracted publication from store");
+
+                        if let Ok(Some((key, publication))) = loaded_element {
+                            // TODO REVIEW: should we be retrying?
+                            // if this failure is due to something that will keep failing it is probably safer to remove and never try again
+                            // otherwise we should retry
+                            debug!("publishing publication {:?} for egress pump", key);
+                            if let Err(e) = publish_handle.publish(publication).await {
+                                error!(message = "failed publishing publication for bridge pump", err = %e);
+                            }
```

Comment:
If this publish fails, what should be our retry logic? I think it depends on why the publish fails. I think the mqtt3 client does some retries by itself. If true then that means this publish might just keep failing.